// machinepack-strings@4.2.0/machines/replace.js
module.exports = {


  friendlyName: 'Replace using regexp',


  description: 'Replace parts of a string that match a given regular expression with the specified replacement.',


  sync: true,


  cacheable: true,


  inputs: {

    string: {
      friendlyName: 'String',
      example: 'Hello world!',
      description: 'The string to modify.',
      required: true
    },

    regexp: {
      friendlyName: 'Regular expression',
      example: 'World',
      description: 'The regular expression to match against (i.e. "metal detector")',
      extendedDescription: 'The regular expression should be specified as a string WIHOUUT including leading or trailing slashes or modifiers like /gi.',
      required: true
    },

    replacement: {
      friendlyName: 'Replacement',
      description: 'The string to use when replacing matches.',
      extendedDescription: 'Note that you can use match expressions (e.g. "$1", "$2", etc.) to express the values of capture groups.',
      example: 'Mumbai',
      required: true
    },

    caseInsensitive: {
      friendlyName: 'Case insensitive?',
      description: 'Whether or not you care about uppercase/lowercase letters.',
      extendedDescription: 'This will build the regular expression using the `/i` modifier.',
      example: true,
      advanced: true,
      defaultsTo: true
    }

  },


  exits: {

    success: {
      friendlyName: 'then',
      description: 'Returns the modified string.',
      example: 'Hello Mumbai!'
    },

    invalidRegexp: {
      friendlyName: 'invalid regexp',
      description: 'Provided regular expression is invalid (cannot be instantiated into a RegExp object)'
    },

    error: {
      description: 'Unexpected error occurred.'
    }

  },


  defaultExit: 'success',


  fn: function (inputs, exits) {

    var _ = require('lodash');

    // Case-insensitive by default
    if (_.isUndefined(inputs.caseInsensitive)) {
      inputs.caseInsensitive = true;
    }

    // Check that the regexp is valid
    var regexp;
    try {

      regexp = inputs.regexp;

      /////////////////////////////////////////////////////////
      // Skip this-- we want users to be able to provide an actual
      // regexp with all the things (i.e. should be able to use the
      // star and dot and ? operators, etc)
      /////////////////////////////////////////////////////////
      // Then escape the provided string before instantiating
      // regexp = _.escapeRegExp(regexp);
      /////////////////////////////////////////////////////////

      // Then construct it
      // (and if relevant, enable case-insensitivity)
      if (inputs.caseInsensitive) {
        regexp = new RegExp(regexp, 'i');
      }
      else {
        regexp = new RegExp(regexp);
      }
    } catch (e) {
      return exits.invalidRegexp(e);
    }

    var newString = inputs.string.replace(regexp, inputs.replacement);

    return exits.success(newString);

  }

};
